File: bindings/e2e/ruby/file_lifecycle.rb
# Public file lifecycle and native archive lock conformance.
require 'tmpdir'
require_relative '../../ruby/revault_api'
ENV['LOCKBOX_LOCK_TIMEOUT_MS'] = '60'
api = Revault.load
signer = api.generate_profile_signing_key_pair
key = 'K' * 32
check = ->(condition) { raise 'file facade assertion failed' unless condition }
refuses = ->(&action) { begin; action.call; rescue Revault::RevaultError; next; end; raise 'operation should have been refused' }
if (archive = ENV['REVAULT_READONLY_ARCHIVE'])
  box = Revault::Lockbox.open(archive, content_key: key)
  begin; check.call(box.get_file('/hello') == 'read only'); ensure; box.close; end
else
  Dir.mktmpdir('revault-ruby-') do |root|
    archive = File.join(root, 'résumé.lbox')
    create = ->(overwrite = false) { Revault::Lockbox.create(archive, content_key: key, signing_key: signer, overwrite: overwrite) }
    open_box = ->(write = false) { Revault::Lockbox.open(archive, content_key: key, signing_key: write ? signer : nil) }
    payload = "hello\0\xff".b
    box = create.call
    begin; box.add_file('/hello', payload, false); box.commit; box.commit; ensure; box.close; end
    reader = open_box.call
    begin
      box = open_box.call
      begin; check.call(box.get_file('/hello') == payload); ensure; box.close; end
      refuses.call { reader.set_owner_signing_key(signer) }
      refuses.call { reader.add_file('/bad', payload, false) }
      refuses.call { open_box.call(true) }
      refuses.call { create.call(true) }
    ensure; reader.close; end
    box = open_box.call(true)
    begin; box.add_file('/hello', 'replacement', true); box.add_file('/added', payload, false); box.commit; ensure; box.close; end
    box = open_box.call
    begin; check.call(box.get_file('/hello') == 'replacement'); check.call(box.get_file('/added') == payload); ensure; box.close; end
    box = open_box.call(true)
    begin; box.delete('/hello'); box.commit; ensure; box.close; end
    box = open_box.call
    begin; check.call(!box.exists('/hello')); check.call(box.get_file('/added') == payload); ensure; box.close; end
    refuses.call { create.call }
    box = create.call(true)
    begin; box.add_file('/new', payload, false); box.commit; ensure; box.close; end
    box = open_box.call
    begin; check.call(!box.exists('/added')); check.call(box.get_file('/new') == payload); ensure; box.close; end
    check.call(Dir.children(root) == ['résumé.lbox'])
    box = Revault::Lockbox.create(archive, password: 'test password', signing_key: signer, overwrite: true)
    begin; box.add_file('/hello', payload, false); box.commit; ensure; box.close; end
    box = Revault::Lockbox.open(archive, password: 'test password')
    begin; check.call(box.get_file('/hello') == payload); ensure; box.close; end
  end
end
signer.free
puts "PASS\truby\tlockbox_file\t#{ENV['REVAULT_READONLY_ARCHIVE'] ? 1 : 18}"

File: bindings/e2e/php/file_lifecycle.php
<?php
/** Public file lifecycle and native archive lock conformance. */
declare(strict_types=1);
require_once __DIR__ . '/../../php/src/DomainModels.php';
require_once __DIR__ . '/../../php/src/BindingOperations.php';
require_once __DIR__ . '/../../php/src/Vault.php';
use Revault\Revault;
use Revault\Lockbox;
use Revault\RevaultError;
putenv('LOCKBOX_LOCK_TIMEOUT_MS=60');
$api = Revault::load();
$signer = $api->generateProfileSigningKeyPair();
$key = str_repeat('K', 32);
$check = static function(bool $condition): void { if (!$condition) throw new RuntimeException('file facade assertion failed'); };
$refuses = static function(callable $action): void { try { $action(); } catch (RevaultError $error) { return; } throw new RuntimeException('operation should have been refused'); };
if (($archive = getenv('REVAULT_READONLY_ARCHIVE')) !== false) {
    $box = Lockbox::open($archive, contentKey: $key);
    try { $check($box->getFile('/hello') === 'read only'); } finally { $box->close(); }
} else {
    $root = sys_get_temp_dir() . '/revault-php-' . bin2hex(random_bytes(8));
    mkdir($root, 0700);
    $archive = $root . '/résumé.lbox';
    $create = static fn(bool $overwrite = false) => Lockbox::create($archive, contentKey: $key, signingKey: $signer, overwrite: $overwrite);
    $open = static fn(bool $write = false) => Lockbox::open($archive, contentKey: $key, signingKey: $write ? $signer : null);
    $payload = "hello\0\xff";
    try {
        $box = $create();
        try { $box->addFile('/hello', $payload, false); $box->commit(); $box->commit(); } finally { $box->close(); }
        $reader = $open();
        try {
            $box = $open();
            try { $check($box->getFile('/hello') === $payload); } finally { $box->close(); }
            $refuses(fn() => $reader->setOwnerSigningKey($signer));
            $refuses(fn() => $reader->addFile('/bad', $payload, false));
            $refuses(fn() => $open(true));
            $refuses(fn() => $create(true));
        } finally { $reader->close(); }
        $box = $open(true);
        try { $box->addFile('/hello', 'replacement', true); $box->addFile('/added', $payload, false); $box->commit(); } finally { $box->close(); }
        $box = $open();
        try { $check($box->getFile('/hello') === 'replacement'); $check($box->getFile('/added') === $payload); } finally { $box->close(); }
        $box = $open(true);
        try { $box->delete('/hello'); $box->commit(); } finally { $box->close(); }
        $box = $open();
        try { $check(!$box->exists('/hello')); $check($box->getFile('/added') === $payload); } finally { $box->close(); }
        $refuses(fn() => $create());
        $box = $create(true);
        try { $box->addFile('/new', $payload, false); $box->commit(); } finally { $box->close(); }
        $box = $open();
        try { $check(!$box->exists('/added')); $check($box->getFile('/new') === $payload); } finally { $box->close(); }
        $check(count(scandir($root)) === 3);
        $box = Lockbox::create($archive, password: 'test password', signingKey: $signer, overwrite: true);
        try { $box->addFile('/hello', $payload, false); $box->commit(); } finally { $box->close(); }
        $box = Lockbox::open($archive, password: 'test password');
        try { $check($box->getFile('/hello') === $payload); } finally { $box->close(); }
    } finally { if (file_exists($archive)) unlink($archive); rmdir($root); }
}
$signer->free();
$assertions = getenv('REVAULT_READONLY_ARCHIVE') !== false ? 1 : 18;
echo "PASS\tphp\tlockbox_file\t$assertions\n";
